File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-option-shared-5-4/Barrier.cpp
// Barrier.cpp 
#include "Barrier.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

Barrier::Barrier(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //barrierCapNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* barrierCapNode = xmlNode->FirstChildElement("barrierCap");

   if(barrierCapNode){barrierCapIsNull_ = false;}
   else{barrierCapIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- barrierCapNode , address : " << barrierCapNode << std::endl;
   #endif
   if(barrierCapNode)
   {
      if(barrierCapNode->Attribute("href") || barrierCapNode->Attribute("id"))
      {
          if(barrierCapNode->Attribute("id"))
          {
             barrierCapIDRef_ = barrierCapNode->Attribute("id");
             barrierCap_ = boost::shared_ptr<TriggerEvent>(new TriggerEvent(barrierCapNode));
             barrierCap_->setID(barrierCapIDRef_);
             IDManager::instance().setID(barrierCapIDRef_,barrierCap_);
          }
          else if(barrierCapNode->Attribute("href")) { barrierCapIDRef_ = barrierCapNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { barrierCap_ = boost::shared_ptr<TriggerEvent>(new TriggerEvent(barrierCapNode));}
   }

   //barrierFloorNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* barrierFloorNode = xmlNode->FirstChildElement("barrierFloor");

   if(barrierFloorNode){barrierFloorIsNull_ = false;}
   else{barrierFloorIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- barrierFloorNode , address : " << barrierFloorNode << std::endl;
   #endif
   if(barrierFloorNode)
   {
      if(barrierFloorNode->Attribute("href") || barrierFloorNode->Attribute("id"))
      {
          if(barrierFloorNode->Attribute("id"))
          {
             barrierFloorIDRef_ = barrierFloorNode->Attribute("id");
             barrierFloor_ = boost::shared_ptr<TriggerEvent>(new TriggerEvent(barrierFloorNode));
             barrierFloor_->setID(barrierFloorIDRef_);
             IDManager::instance().setID(barrierFloorIDRef_,barrierFloor_);
          }
          else if(barrierFloorNode->Attribute("href")) { barrierFloorIDRef_ = barrierFloorNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { barrierFloor_ = boost::shared_ptr<TriggerEvent>(new TriggerEvent(barrierFloorNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<TriggerEvent> Barrier::getBarrierCap()
{
   if(!this->barrierCapIsNull_){
        if(barrierCapIDRef_ != ""){
             return boost::shared_static_cast<TriggerEvent>(IDManager::instance().getID(barrierCapIDRef_));
        }else{
             return this->barrierCap_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<TriggerEvent>();
   }
}
boost::shared_ptr<TriggerEvent> Barrier::getBarrierFloor()
{
   if(!this->barrierFloorIsNull_){
        if(barrierFloorIDRef_ != ""){
             return boost::shared_static_cast<TriggerEvent>(IDManager::instance().getID(barrierFloorIDRef_));
        }else{
             return this->barrierFloor_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<TriggerEvent>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-option-shared-5-4/Barrier.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class Barrier : ISerialized
    {
        public Barrier(XmlNode xmlNode)
        {
            XmlNode barrierCapNode = xmlNode.SelectSingleNode("barrierCap");
            
            if (barrierCapNode != null)
            {
                if (barrierCapNode.Attributes["href"] != null || barrierCapNode.Attributes["id"] != null) 
                {
                    if (barrierCapNode.Attributes["id"] != null) 
                    {
                        barrierCapIDRef_ = barrierCapNode.Attributes["id"].Value;
                        TriggerEvent ob = new TriggerEvent(barrierCapNode);
                        IDManager.SetID(barrierCapIDRef_, ob);
                    }
                    else if (barrierCapNode.Attributes["href"] != null)
                    {
                        barrierCapIDRef_ = barrierCapNode.Attributes["href"].Value;
                    }
                    else
                    {
                        barrierCap_ = new TriggerEvent(barrierCapNode);
                    }
                }
                else
                {
                    barrierCap_ = new TriggerEvent(barrierCapNode);
                }
            }
            
        
            XmlNode barrierFloorNode = xmlNode.SelectSingleNode("barrierFloor");
            
            if (barrierFloorNode != null)
            {
                if (barrierFloorNode.Attributes["href"] != null || barrierFloorNode.Attributes["id"] != null) 
                {
                    if (barrierFloorNode.Attributes["id"] != null) 
                    {
                        barrierFloorIDRef_ = barrierFloorNode.Attributes["id"].Value;
                        TriggerEvent ob = new TriggerEvent(barrierFloorNode);
                        IDManager.SetID(barrierFloorIDRef_, ob);
                    }
                    else if (barrierFloorNode.Attributes["href"] != null)
                    {
                        barrierFloorIDRef_ = barrierFloorNode.Attributes["href"].Value;
                    }
                    else
                    {
                        barrierFloor_ = new TriggerEvent(barrierFloorNode);
                    }
                }
                else
                {
                    barrierFloor_ = new TriggerEvent(barrierFloorNode);
                }
            }
            
        
        }
        
    
        #region BarrierCap_
        private TriggerEvent barrierCap_;
        public TriggerEvent BarrierCap_
        {
            get
            {
                if (this.barrierCap_ != null)
                {
                    return this.barrierCap_; 
                }
                else if (this.barrierCapIDRef_ != null)
                {
                    barrierCap_ = IDManager.getID(barrierCapIDRef_) as TriggerEvent;
                    return this.barrierCap_; 
                }
                else
                {
                      return this.barrierCap_; 
                }
            }
            set
            {
                if (this.barrierCap_ != value)
                {
                    this.barrierCap_ = value;
                }
            }
        }
        #endregion
        
        public string barrierCapIDRef_ { get; set; }
        #region BarrierFloor_
        private TriggerEvent barrierFloor_;
        public TriggerEvent BarrierFloor_
        {
            get
            {
                if (this.barrierFloor_ != null)
                {
                    return this.barrierFloor_; 
                }
                else if (this.barrierFloorIDRef_ != null)
                {
                    barrierFloor_ = IDManager.getID(barrierFloorIDRef_) as TriggerEvent;
                    return this.barrierFloor_; 
                }
                else
                {
                      return this.barrierFloor_; 
                }
            }
            set
            {
                if (this.barrierFloor_ != value)
                {
                    this.barrierFloor_ = value;
                }
            }
        }
        #endregion
        
        public string barrierFloorIDRef_ { get; set; }
        
    
        
    
    }
    
}

